### ISSUE TITLE
Running gatsby develop when gatsby-source-contentful configured to work with "Preview API" does not delete Gatsby nodes when entries are deleted from Contentful

### ISSUE THREAD

**Issue by smnh** [gatsbyjs/gatsby#23415]
## Description

When `gatsby-source-contentful` is [configured to work with the Preview API](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/README.md#using-preview-api), this plugin does not delete Gatsby nodes when entries are deleted from Contentful.

```
module.exports = {
  plugins: [
    {
      resolve: `gatsby-source-contentful`,
      options: {
        spaceId: `...`,
        accessToken: process.env.CONTENTFUL_ACCESS_TOKEN,
        host: `preview.contentful.com`,
      },
    },
  ],
}
```

### Steps to reproduce

- Configure `gatsby-source-contentful` with `host: 'preview.contentful.com'`
- Run `gatsby develop`
- Add an entry to Contentful space
- Restart gatsby develop
- The new entry appears in GraphQL (and on a page if it renders entries of that type)
- Delete that entry from Contentful space
- Restart gatsby develop
- The entry still appears in GraphQL and on a page

### Expected result

When an entry is deleted from Contentful, its respective Gatsby node should be deleted as well and should not appear in GraphQL (or a rendered page).

### Actual result

When an entry is deleted from Contentful, its respective Gatsby node is not deleted, and therefore continue appearing in GraphQL (or a rendered page).

### Notes

This issue can be temporary solved by cleaning the cache every time remote content is changed, for example by running `gatsby clean`. But this makes the development process cumbersome and prevent utilization of cache mechanisms.

This issue doesn't occur when `gatsby-source-contentful` configured to work with Contentful Delivery API (`host: 'cdn.contentful.com'`).

The underlying problem is that `gatsby-source-contentful` uses Contentful's [Synchronization API](https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/synchronization) with Contentful's Preview API. While Contentful [specify](https://www.contentful.com/developers/docs/references/content-preview-api/#/introduction/preview-api-limitations) that Synchronization API is not supported with Preview API:

> The Preview API does not implement the Sync API, so applications that rely exclusively on the Sync API to load data will not be usable with the Preview API.

Yet, although Contentful own specification, some aspects of Synchronization API do work with Preview API. Otherwise, the content wouldn't be fetched at all. But it seems that some of its aspects doesn't work. Such as when entries are removed, the sync API does not say that. Here are the refined steps to reproduce the underlying problem:

1. Call [http://cdn.contentful.com/spaces/:spaceId/environments/:environment/sync?initial=true](http://cdn.contentful.com/spaces/:spaceId/environments/:environment/sync?initial=true)
    * continue calling the returned `nextSyncUrl` until no `items` are returned
2. Call [http://preview.contentful.com/spaces/:spaceId/environments/:environment/sync?initial=true](http://preview.contentful.com/spaces/:spaceId/environments/:environment/sync?initial=true)
    * continue calling the returned `nextSyncUrl` until no `items` are returned
3. Add an entry to Contentful space, and publish it
4. Call [http://preview.contentful.com/spaces/:spaceId/environments/:environment/sync?sync_token=](http://preview.contentful.com/spaces/:spaceId/environments/:environment/sync?sync_token=)...
    * you should get back `items` with the single entry that was just created and published
    * call this endpoint again with the returned `nextSyncUrl`, you should get back empty `items` list
5. Call [http://cdn.contentful.com/spaces/:spaceId/environments/:environment/sync?sync_token=](http://cdn.contentful.com/spaces/:spaceId/environments/:environment/sync?sync_token=)...
    * you should get back `items` with the single entry that was just created and published.
    * call this endpoint again with the returned `nextSyncUrl`, you should get back empty `items` list
6. delete the entry from Contentful
7. call these two endpoints again
8. **Only [cdn.contentful.com](http://cdn.contentful.com) API returns the deleted entry, the [preview.contentful.com](http://preview.contentful.com) API does not!**

cdn.contentful.com response with the deleted entry:
```
{
  "sys": {
    "type": "Array"
  },
  "items": [
    {
      "sys": {
        "type": "DeletedEntry",
        "id": "...",
        "space": { ... },
        "environment": { ... }
        },
        "revision": 1,
        "createdAt": "...",
        "updatedAt": "...",
        "deletedAt": "..."
      }
    }
  ],
  "nextSyncUrl": "https://cdn.contentful.com/spaces/.../environments/.../sync?sync_token=..."
}
```

preview.contentful.com response without the deleted entry:
```
{
  "sys": {
    "type": "Array"
  },
  "items": [],
  "nextSyncUrl": "http://preview.contentful.com/spaces/52sie3aajp6k/environments/staging/sync?sync_token=..."
}
```

### Environment

Configure `gatsby-source-contentful` to use Preview API

```
    {
      resolve: `gatsby-source-contentful`,
      options: {
        spaceId: `...`,
        accessToken: process.env.CONTENTFUL_ACCESS_TOKEN,
        host: `preview.contentful.com`,
      },
```

Run `gatsby develop`

**Comment by pieh** [comment]
First of all - thanks for doing lot of research on this and providing very detailed issue description!

I'm not sure how we could approach this to be honest: while we could try to use non-sync APIs for preview, this will cause perf degradation for larger Contentful spaces (doing full sync on each change will be pretty expensive) and there will be trade off needed to be made: accept limitation (not deleting nodes) or accept perf degradation.

I will think about this more and also try to chat with Contentful if this limitation of Preview API could potentially be lifted

**Comment by pieh** [comment]
As a workaround - can you try conditionally using `forceFullSync: true` plugin option (when you are using preview endpoint)? I'm not sure if it will fix the issue TBH, but it's worth a try

**Comment by axe312ger** [comment]
This might be fixed as soon #25249 is published

**Comment by LekoArts** [comment]
Closing this as stale since in the meantime Gatsby v3 and updated related packages were released. Please try with the latest versions and if you still see this problem open a new bug report (it must include a [minimal reproduction][reproduction]).

[reproduction]: https://gatsby.dev/reproduction
